server/services/v1/database/secondary_indexer.go

```diff
@@ -134,6 +134,105 @@ func NewSecondaryIndexer(coll *schema.DefaultCollection) *SecondaryIndexer {
 	}
 }
 
+func (q *SecondaryIndexer) BuildCollection(ctx context.Context, txMgr *transaction.Manager) error {
+	docFetch := 500
+	var last []byte
+	var first []byte
+	count := 0
+	batchCount := 0
+
+	for {
+		tx, err := txMgr.StartTx(ctx)
+		if err != nil {
+			return err
+		}
+
+		iter, err := createBulkDocsReader(ctx, tx, q.coll.EncodedName, first, last)
+		if err != nil {
+			return err
+		}
+
+		var row Row
+		for iter.Next(&row) && count <= docFetch {
+			// Record first key so that if this batch fails
+			// we know which key to restart from
+			if count == 0 {
+				first = row.Key
+			}
+
+			last = row.Key
+			fdbKey, err := keys.FromBinary(q.coll.EncodedName, row.Key)
+			if err != nil {
+				return err
+			}
+
+			if err = q.Index(ctx, tx, row.Data, fdbKey.IndexParts()); err != nil {
+				return err
+			}
+			count += 1
+		}
+
+		err = tx.Commit(ctx)
```

Comment:
nit, you can combine the below `if` in this line i.e.
```
if err = tx.Commit(ctx); err != nil {
  if !shouldRetryBulkIndex(err) {
    return err
  }
  
  docFetch /= 2
  count = 0
  continue
}
```

Reply:
Nice catch